### PR TITLE
Verify now allows missing genotypes

### LINKED ISSUE #625: tsinfer verify does not take missing data into account
At a minimum, we should detect the existance of missing data in the input and throw an error. A bit better would be to only match the non-missing genotypes.

### Discussed in https://github.com/tskit-dev/tsinfer/discussions/624

<div type='discussions-op-text'>

<sup>Originally posted by **chlangley** January 18, 2022</sup>
Hello:
I have stumbled around this early impasse long enough.  Time for a clue, please.

While I have gotten my data into a valid Sample and been able to add metadata (sampleIDs and populationIDs) and have run tsinfer infer on both the entire chromosome and on a small sub regions, test_direct.samples (see attached [tsinfer_infer_out.txt](https://github.com/tskit-dev/tskit/files/7885121/tsinfer_infer_out.txt).  Note the unnerving line 13 in 'tsinfer_infer_out.txt' where the progress begins unexpectedly: "... sites; 7301 mutations######## ...".

When I examine indivdual trees on a simplify(d) subsample they appear to fit my expectations from previous analyses of these data.  

Yet the resulting test_direct.trees does not tsinfer verify with test_direct.samples (see [tsinfer_verify_out.txt](https://github.com/tskit-dev/tskit/files/7885127/tsinfer_verify_out.txt) .  This seems like a good reason to pause and find out the meaning of the deviant genotypes.

The input file looks like this in the upper lefthand corner (snippet) where one might expect the " Alleles for sample 6 not equal at site 4" to be.
`| => less ../cpragts/kgp4_chrX_hap_ba_snp`
`2781487 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781514 0 0 1 0 1 1 0 0 0 1 0 0 0 0 1 0 0`
`2781518 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781584 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0`
`2781603 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781604 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781623 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781635 1 0 0 0 0 0 0 0 1 0 1 0 1 0 0 1 0`
`2781642 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781670 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781698 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781865 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781878 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781893 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`
`2781927 0 1 1 1 1 0 1 0 0 0 0 1 0 1 0 0 1`
`2781983 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`

 I don't see an obvious candidate.  Everything is 0,1,2 or -1.
How might one proceed to determine the source of this error?  

thanks for any help.
Cheers,
Chuck

</div>

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #625

### CODE REVIEW

**Comment by benjeffery** [comment]
@jeromekelleher I think I've got the right end of the stick here?

**Review by jeromekelleher — COMMENTED** [review]
This is great, thanks @benjeffery. I think we can/probably should simplify the logic a bit as we can't have missing data in tsinfer'd output tree sequences.

File: tests/test_inference.py [hunk 1]
```diff
@@ -2901,6 +2901,28 @@ def test_nominal_case(self):
...
         tsinfer.verify(samples, ts)
 
+    def test_missingness(self):
+        ts = msprime.simulate(10, mutation_rate=5, random_seed=1)
+        assert ts.num_sites > 0
+        samples = tsinfer.SampleData.from_tree_sequence(ts)
+        inferred_ts = tsinfer.infer(samples)
+        tsinfer.verify(samples, inferred_ts)
+
+        # Mark some of the samples as missing in the ts by removing the edges that
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply tests/test_inference.py:2908]: I've change the test so the ts with missingness is used to create the sample data, but the sample data genotypes have no missingness in - is this expected?

[the remark this answers is not in the document]
Comment (benjeffery) [reply tests/test_inference.py:2908]: Ah-ha! The error was in the code that checks for missingness in the array. I needed to convert the zarr array to a numpy one.

**Comment by benjeffery** [comment]
I've (hopefully) addressed the comments and added a changelog.

**Review by jeromekelleher — APPROVED** [review]

**Pull request merged**